### ISSUE TITLE
Micrometer client not reconnecting to statsd after server is restarted

### ISSUE THREAD

**Issue by dunnk2022** [micrometer-metrics/micrometer#3563]
**Describe the bug**
The micrometer client does not reconnect to the statsd server (Datadog agent) when the agent is restarted.

**Environment**

 - Micrometer version - 1.9.5
 - Micrometer registry - statsd, Datadog flavor 
 - OS: Linux
 - Java version: 11.0.14

**To Reproduce**
How to reproduce the bug:
Restart the statsd server (datadog-agent)

**Expected behavior**

The micrometer client re-establishes a connection to the statsd server on port 8125 when the daemon is back up.

**Additional context**
Add any other context about the problem here, e.g. related issues.

**Comment by dunnk2022** [comment]
Thanks for the response @shakuzen

**Comment by dunnk2022** [comment]
We really just do a rolling restart of the agent and the issue occurs. Perhaps turning DEBUG logging on for the client will help to gather more info?

**Comment by shakuzen** [comment]
If you turn on debug logging for classes under the `io.micrometer.shaded.reactor.netty` package, we can probably get some more information about what's happening. Or turn on debug logging for everything under `io.micrometer`.

**Comment by shakuzen** [comment]
@raymondchen625 thanks for the additional datapoint. Unfortunately, since I haven't been able to reproduce it locally or get logs from anyone with more info about what is happening, we haven't made any progress on this. If you would be able to provide steps to reproduce this (ideally locally or with as minimal other things involved as possible) or some debug logs, it may help get to the bottom of this.

**Comment by raymondchen625** [comment]
> @raymondchen625 thanks for the additional datapoint. Unfortunately, since I haven't been able to reproduce it locally or get logs from anyone with more info about what is happening, we haven't made any progress on this. If you would be able to provide steps to reproduce this (ideally locally or with as minimal other things involved as possible) or some debug logs, it may help get to the bottom of this.

I could not reproduce this locally. Unfortunately, I didn't try to capture IP packets before restarting the pods to see if Micrometer was sending traffic somewhere. I did check the application logs and didn't see any error. For custom metrics being sent to UDP 8125, I believe there is no error logs if the agent is down. But we also had APM metrics sent to TCP 8126, which should print error logs if the service is down. That makes me think Micrometer might have stopped sending out metrics. I might have captured nothing even if I had done the tcpdump on it.

**Comment by shakuzen** [comment]
I suspect (but could be wrong) that there will be some logs at debug level when the issue occurs, but most people don't have debug logs enabled on their production services.

**Comment by raymondchen625** [comment]
> I suspect (but could be wrong) that there will be some logs at debug level when the issue occurs, but most people don't have debug logs enabled on their production services.

We found the same issue in our staging environment. I killed one of our two pods now it's clearly from the Datadog dashboard that only the new pod is able to send metrics. I used tcpdump to capture IP packets on both containers. They are both sending TCP packets to port 8126 or UDP datagrams to port 8125.
It seems there is something wrong on the Datadog agent side. When I further investigate it, I would set the log level of `io.micrometer.shaded.reactor.netty` to get better insight.

**Comment by raymondchen625** [comment]
Hi @shakuzen, I finally somehow found a scenario to reproduce the issue. 
There are a few facts about `micrometer-registry-statsd` component (`1.9.17` in my test):
1. It doesn't print out any logging message when there are problems sending data to the UDP port of the target host, whether it's an unresolvable hostname or the statsd service on the target host is down. It's just totally quiet.
2. The IP resolution only happens once when the `UdpClient` instance is initialized. The instance is reused and no further DNS resolution will happen.

To reproduce the issue:
1. Put the statsd services listening on a UDP port behind a Kubernetes service. And configure the client application's `statsdHost` to the service's [Kubernetes internal DNS record name](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#dns-records), such as `mystatsd.mynamespace.svc.cluster.local`
2. Deploy both the statsd service and the test application. We should be able to see the UDP datagrams are routed to one of the pods behind the statsd service.
3. Undeploy the statsd service. No error log will be printed out from the client application, If we log into the client application's container and use `netstat` or `lsof` to quey the UDP connections, we should be able to see the socket whose destination IP is still the old service hostname's former IP address, which should not work anymore by now.
4. Redeploy the test statsd Kubernetes service, we will see the hostname `mystatsd.mynamespace.svc.cluster.local` being resolved to a new IP. But the client application is still sending datagraes to an non-existent old IP.

I'm not sure if this is exactly the same scenario @dunnk2022 had. But this indicates an issue with the current implementation: it doesn't resolve the hostname again after `UdpClient` initialization.
Ideally, it should be able to be configured to detect DNS resolution changes and reinitialize `UdpClient` to send metrics to the new destination. If that's not possible, at least print some error logs.

**Comment by dunnk2022** [comment]
@raymondchen625 We added an initContainer to our Datadog host agent pods to delete conntrack entries before starting the Datadog agent.

This change was based on comments from this issue - https://github.com/containernetworking/plugins/issues/123.

We have not had a report of missing metrics sent over StatsD since.

**Comment by marcingrzejszczak** [comment]
After looking at the latest comments, can we assume that the problem is fixed ?

**Comment by shakuzen** [comment]
@raymondchen625 What you've described sounds like https://github.com/micrometer-metrics/micrometer/issues/1252. If that is still an issue in the latest versions, please comment over there so we know it still needs to be fixed. I'm not sure if that was the original problem here or not. If the server was restarted with the same IP, I guess it's a different problem.

**Comment by raymondchen625** [comment]
> If the server was restarted with the same IP, I guess it's a different problem.

Yes, that's probably related to a series of kube-proxy issues:
- https://github.com/kubernetes/kubernetes/issues/101607
- https://github.com/kubernetes/kubernetes/pull/119394
- https://github.com/kubernetes/kubernetes/pull/104009

IMO, this issue can be closed. The DNS change detection issue can be addressed by #1252 .

**Comment by marcingrzejszczak** [comment]
Closing in favour of https://github.com/micrometer-metrics/micrometer/issues/1252

**Comment by dunnk2022** [comment]
I would say that my reported issue is fixed.

Thanks All!

**Comment by marcingrzejszczak** [comment]
@dunnk2022 how is it fixed? I'm asking cause https://github.com/micrometer-metrics/micrometer/issues/1252 seems to be opened

**Comment by dunnk2022** [comment]
@marcingrzejszczak My issue was not on the micrometer end. 
It is related to stale conntrack entries and the portmap version used within our CNI (Flannel v0.11) as described here - https://github.com/containernetworking/plugins/issues/123.


> @raymondchen625 We added an initContainer to our Datadog host agent pods to delete conntrack entries before starting the Datadog agent.
> 
> This change was based on comments from this issue - [containernetworking/plugins#123](https://github.com/containernetworking/plugins/issues/123).
> 
> We have not had a report of missing metrics sent over StatsD since.
